### PR TITLE
GG-18672: [ML] PeerClassloading for ml related lambdas

### PR DESCRIPTION
**Pull request by avplatonov**

### CODE REVIEW

File: modules/ml/src/main/java/org/apache/ignite/ml/trainers/DatasetTrainer.java [hunk 1]
```diff
@@ -85,7 +85,36 @@
      * @param <V> Type of a value in {@code upstream} data.
      * @return Model.
      */
-    public abstract <K, V> M fit(DatasetBuilder<K, V> datasetBuilder, Preprocessor<K, V> preprocessor);
+    public <K, V> M fit(DatasetBuilder<K, V> datasetBuilder, Preprocessor<K, V> preprocessor) {
+        learningEnvironment().deployContext().init(preprocessor);
```
Comment (dmitrievanthony) [inline modules/ml/src/main/java/org/apache/ignite/ml/trainers/DatasetTrainer.java]: Does it make sense to add method `initDeployContext` into the learning environment and don't expose deploy context object itself?

File: modules/ml/src/main/java/org/apache/ignite/ml/trainers/DatasetTrainer.java [hunk 2]
```diff
@@ -85,7 +85,36 @@
...
+
+        return fit0(datasetBuilder, preprocessor);
+    }
+
+    /**
+     * Trains model based on the specified data.
+     *
+     * @param datasetBuilder Dataset builder.
+     * @param preprocessor Extractor of {@link UpstreamEntry} into {@link LabeledVector}.
```
Comment (dmitrievanthony) [inline modules/ml/src/main/java/org/apache/ignite/ml/trainers/DatasetTrainer.java]: One parameter (`learningEnvironment`) is missed.

File: modules/ml/src/test/java/org/apache/ignite/ml/util/generators/DataStreamGeneratorFillCacheTest.java [hunk 3]
```diff
@@ -29,40 +32,28 @@
...
-
     /** */
     @Test
     public void testCacheFilling() {
-        IgniteConfiguration configuration = new IgniteConfiguration().setPeerClassLoadingEnabled(true);
+        IgniteConfiguration configuration = new IgniteConfiguration()
+            .setDiscoverySpi(new TcpDiscoverySpi()
+                .setIpFinder(new TcpDiscoveryVmIpFinder()
+                    .setAddresses(Arrays.asList("127.0.0.1:47500..47509"))))
+            .setPeerClassLoadingEnabled(true);
```
Comment (dmitrievanthony) [inline modules/ml/src/test/java/org/apache/ignite/ml/util/generators/DataStreamGeneratorFillCacheTest.java]: Looks like it doesn't work in tests.

File: modules/tensorflow/src/main/java/org/apache/ignite/tensorflow/cluster/spec/TensorFlowServerAddressSpec.java [hunk 4]
```diff
@@ -18,6 +18,8 @@
 
 import java.io.Serializable;
 import java.util.Collection;
+import java.util.HashMap;
```
Comment (dmitrievanthony) [inline modules/tensorflow/src/main/java/org/apache/ignite/tensorflow/cluster/spec/TensorFlowServerAddressSpec.java]: Please revert this file.

File: examples/src/main/java/org/apache/ignite/examples/ml/dataset/AlgorithmSpecificDatasetExample.java [hunk 5]
```diff
@@ -86,24 +87,26 @@ public static void main(String[] args) throws Exception {
...
+                IgniteBiFunction<SimpleLabeledDatasetData, AlgorithmSpecificPartitionContext, SimpleLabeledDatasetData> builderFun = (data, ctx) -> {
+                    double[] features = data.getFeatures();
+                    int rows = data.getRows();
+
+                    // Makes a copy of features to supplement it by columns with values equal to 1.0.
+                    double[] a = new double[features.length + rows];
 
-                        // Makes a copy of features to supplement it by columns with values equal to 1.0.
-                        double[] a = new double[features.length + rows];
+                    for (int i = 0; i < rows; i++)
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply examples/src/main/java/org/apache/ignite/examples/ml/dataset/AlgorithmSpecificDatasetExample.java]: +

File: examples/src/main/java/org/apache/ignite/examples/ml/regression/linear/LinearRegressionLSQRTrainerExample.java [hunk 6]
```diff
@@ -27,6 +27,7 @@
 import org.apache.ignite.ml.regressions.linear.LinearRegressionModel;
 import org.apache.ignite.ml.selection.scoring.evaluator.Evaluator;
 import org.apache.ignite.ml.selection.scoring.metric.regression.RegressionMetrics;
+import org.apache.ignite.ml.structures.LabeledVector;
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply examples/src/main/java/org/apache/ignite/examples/ml/regression/linear/LinearRegressionLSQRTrainerExample.java]: ok

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java [hunk 7]
```diff
@@ -102,7 +103,7 @@ public GmmTrainer(int countOfComponents) {
     }
 
     /** {@inheritDoc} */
-    @Override public <K, V> GmmModel fit(DatasetBuilder<K, V> datasetBuilder,
+    @Override public <K, V> GmmModel fit0(DatasetBuilder<K, V> datasetBuilder,
```
[the remark this answers is not in the document]
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/clustering/gmm/GmmTrainer.java]: ok, I renamed it to 'fitWithInitializedDeployingContetx'

File: examples/src/main/java/org/apache/ignite/examples/ml/multiclass/OneVsRestClassificationExample.java [hunk 8]
```diff
@@ -25,6 +25,7 @@
 import org.apache.ignite.Ignition;
 import org.apache.ignite.cache.query.QueryCursor;
 import org.apache.ignite.cache.query.ScanQuery;
+import org.apache.ignite.ml.dataset.feature.extractor.Vectorizer;
```
Comment (ybabak) [inline examples/src/main/java/org/apache/ignite/examples/ml/multiclass/OneVsRestClassificationExample.java]: Unnecessary import

File: modules/ml/src/main/java/org/apache/ignite/ml/clustering/kmeans/KMeansModel.java [hunk 9]
```diff
@@ -29,7 +31,9 @@
 /**
  * This class encapsulates result of clusterization by KMeans algorithm.
  */
-public class KMeansModel implements ClusterizationModel<Vector, Integer>, Exportable<KMeansModelFormat> {
+public final class KMeansModel implements ClusterizationModel<Vector, Integer>, Exportable<KMeansModelFormat>,
+    DeployableObject {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/clustering/kmeans/KMeansModel.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/bagging/BaggedTrainer.java [hunk 10]
```diff
@@ -156,6 +156,9 @@ public BaggedTrainer(DatasetTrainer<? extends IgniteModel, L> tr,
     /** {@inheritDoc} */
     @Override public <K, V> BaggedModel update(BaggedModel mdl, DatasetBuilder<K, V> datasetBuilder,
         Preprocessor<K, V> preprocessor) {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/bagging/BaggedTrainer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/convergence/ConvergenceChecker.java [hunk 11]
```diff
@@ -90,10 +91,15 @@ public boolean isConverged(
         LearningEnvironmentBuilder envBuilder,
         DatasetBuilder<K, V> datasetBuilder,
         ModelsComposition currMdl) {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/convergence/ConvergenceChecker.java]: Unnecessary empty line

Comment (ybabak) [reply modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/convergence/ConvergenceChecker.java]: same for line 72

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java [hunk 12]
```diff
@@ -68,20 +68,23 @@ public GDBBinaryClassifierTrainer(double gradStepSize, Integer cntOfIterations,
     @Override protected <V, K> boolean learnLabels(DatasetBuilder<K, V> builder,
         Preprocessor<K, V> preprocessor) {
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBBinaryClassifierTrainer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java [hunk 13]
```diff
@@ -186,10 +186,13 @@ public GDBTrainer(double gradStepSize, Integer cntOfIterations, Loss loss) {
         DatasetBuilder<K, V> builder,
         Preprocessor<K, V> preprocessor) {
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/boosting/GDBTrainer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/combinators/parallel/TrainersParallelComposition.java [hunk 14]
```diff
@@ -94,6 +94,9 @@
     @Override public <K, V> IgniteModel<I, List<O>> update(IgniteModel<I, List<O>> mdl,
         DatasetBuilder<K, V> datasetBuilder,
         Preprocessor<K, V> preprocessor) {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/combinators/parallel/TrainersParallelComposition.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/stacking/StackedDatasetTrainer.java [hunk 15]
```diff
@@ -239,6 +239,7 @@ public StackedDatasetTrainer() {
     @Override public <K, V> StackedModel<IS, IA, O, AM> update(StackedModel<IS, IA, O, AM> mdl,
         DatasetBuilder<K, V> datasetBuilder, Preprocessor<K, V> preprocessor) {
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/stacking/StackedDatasetTrainer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/ModelsComposition.java [hunk 16]
```diff
@@ -18,17 +18,21 @@
...
 import org.apache.ignite.ml.math.primitives.vector.Vector;
 import org.apache.ignite.ml.util.ModelTrace;
 
 /**
  * Model consisting of several models and prediction aggregation strategy.
  */
-public class ModelsComposition implements IgniteModel<Vector, Double>, Exportable<ModelsCompositionFormat> {
+public class ModelsComposition implements IgniteModel<Vector, Double>, Exportable<ModelsCompositionFormat>,
+    DeployableObject {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/ModelsComposition.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/dataset/feature/extractor/impl/BinaryObjectVectorizer.java [hunk 17]
```diff
@@ -31,7 +33,9 @@
  *
  * @param <K> Type of key.
  */
-public class BinaryObjectVectorizer<K> extends ExtractionUtils.StringCoordVectorizer<K, BinaryObject> {
+public final class BinaryObjectVectorizer<K> extends ExtractionUtils.StringCoordVectorizer<K, BinaryObject>
+    implements DeployableObject {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/dataset/feature/extractor/impl/BinaryObjectVectorizer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/dataset/impl/cache/util/ComputeUtils.java [hunk 18]
```diff
@@ -120,6 +127,49 @@
         throw new IllegalStateException();
     }
 
+    /**
+     * Callable that contains deploy context and can pass missing classes
+     * during learning session by p2p deployment.
+     * @param <C> Type of callable result.
+     */
+    private static class DeployableCallable<C> implements GridPeerDeployAware, IgniteCallable<C> {
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/dataset/impl/cache/util/ComputeUtils.java]: please, reorder methods and inner class

File: modules/ml/src/main/java/org/apache/ignite/ml/pipeline/PipelineMdl.java [hunk 19]
```diff
@@ -66,4 +68,8 @@
             "internalMdl=" + internalMdl +
             '}';
     }
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/pipeline/PipelineMdl.java]: missed javadoc

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderPreprocessor.java [hunk 20]
```diff
@@ -17,7 +17,9 @@
 package org.apache.ignite.ml.preprocessing.encoding;
 
 import java.util.Map;
+import java.util.Optional;
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/encoding/EncoderPreprocessor.java]: Unnecessary imports

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/maxabsscaling/MaxAbsScalerTrainer.java [hunk 21]
```diff
@@ -37,6 +38,7 @@
         LearningEnvironmentBuilder envBuilder,
         DatasetBuilder<K, V> datasetBuilder,
         Preprocessor<K, V> basePreprocessor) {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/maxabsscaling/MaxAbsScalerTrainer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/maxabsscaling/MaxAbsScalerTrainer.java [hunk 22]
```diff
@@ -20,6 +20,7 @@
 import org.apache.ignite.ml.dataset.DatasetBuilder;
 import org.apache.ignite.ml.dataset.UpstreamEntry;
 import org.apache.ignite.ml.dataset.primitive.context.EmptyContext;
+import org.apache.ignite.ml.environment.LearningEnvironment;
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/maxabsscaling/MaxAbsScalerTrainer.java]: Unnecessary import

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/minmaxscaling/MinMaxScalerTrainer.java [hunk 23]
```diff
@@ -38,6 +38,7 @@
         LearningEnvironmentBuilder envBuilder,
         DatasetBuilder<K, V> datasetBuilder,
         Preprocessor<K, V> basePreprocessor) {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/minmaxscaling/MinMaxScalerTrainer.java]: Unnecessary empty line

File: modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/standardscaling/StandardScalerTrainer.java [hunk 24]
```diff
@@ -55,6 +55,7 @@
     private StandardScalerData computeSum(LearningEnvironmentBuilder envBuilder,
         DatasetBuilder<K, V> datasetBuilder,
         Preprocessor<K, V>  basePreprocessor) {
+
```
Comment (ybabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/preprocessing/standardscaling/StandardScalerTrainer.java]: Unnecessary empty line

**Pull request merged**